src/lib/server/server.cpp

```diff
@@ -1,40 +1,55 @@
 #include "server.hpp"
 
-#include <boost/asio/placeholders.hpp>
-#include <boost/bind.hpp>
+#include <pthread.h>
 
-#include "client_connection.hpp"
-#include "server_session.hpp"
-#include "task_runner.hpp"
-#include "then_operator.hpp"
+#include <iostream>
+#include <thread>
 
 namespace opossum {
 
-using opossum::then_operator::then;
-
-Server::Server(boost::asio::io_service& io_service, uint16_t port)
-    : _io_service(io_service),
-      _acceptor(io_service, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), port)),
-      _socket(io_service) {
-  _accept_next_connection();
+// Specified port (default: 5432) will be opened after initializing the _acceptor
+Server::Server(const boost::asio::ip::address& address, const uint16_t port,
+               const SendExecutionInfo send_execution_info)
+    : _acceptor(_io_service, boost::asio::ip::tcp::endpoint(address, port)), _send_execution_info(send_execution_info) {
+  std::cout << "Server started at " << server_address() << " and port " << server_port() << std::endl
+            << "Run 'psql -h localhost' to connect to the server" << std::endl;
 }
 
-void Server::_accept_next_connection() {
-  _acceptor.async_accept(_socket, boost::bind(&Server::_start_session, this, boost::asio::placeholders::error));
+void Server::run() {
+  _accept_new_session();
+  _io_service.run();
 }
 
-void Server::_start_session(boost::system::error_code error) {
-  if (!error) {
-    auto connection = std::make_shared<ClientConnection>(std::move(_socket));
-    auto task_runner = std::make_shared<TaskRunner>(_io_service);
-    auto session = std::make_shared<ServerSession>(connection, task_runner);
-    // Start the session and release it once it has terminated
-    session->start() >> then >> [=]() mutable { session.reset(); };
-  }
+void Server::_accept_new_session() {
+  // Create a new session. This will also open a new data socket in order to communicate with the client
+  // For more information on TCP ports + Asio see:
+  // https://www.gamedev.net/forums/topic/586557-boostasio-allowing-multiple-connections-to-a-single-server-socket/
+  auto new_session = std::make_shared<Session>(_io_service, _send_execution_info);
+  _acceptor.async_accept(*(new_session->socket()),
+                         boost::bind(&Server::_start_session, this, new_session, boost::asio::placeholders::error));
+}
 
-  _accept_next_connection();
+void Server::_start_session(const std::shared_ptr<Session>& new_session, const boost::system::error_code& error) {
+  Assert(!error, error.message());
+
+  std::thread session_thread([new_session] {
+    const std::string thread_name = "port_" + std::to_string(new_session->socket()->remote_endpoint().port());
```

Comment:
Maybe, server_port to make more explicit from which component the thread comes from.

Reply:
According to the [documentation](http://man7.org/linux/man-pages/man3/pthread_setname_np.3.html) of `pthread_setname_np` (linux) the length is restricted to 16 chars including the null terminator.

Reply:
Then, maybe server_pXXXXX (my preference) or server_p_XXXXX? I believe that there is value in fully stating the component for debugging.

Reply:
server_p_XXXXX should be fine.